alternator/executor.cc

```diff
@@ -1297,30 +1297,32 @@ class put_item_operation : public rmw_operation {
 };
 
 future<executor::request_return_type> executor::put_item(client_state& client_state, std::string content) {
-    _stats.api_operations.put_item++;
-    auto start_time = std::chrono::steady_clock::now();
-    rjson::value request = rjson::parse(content);
-    elogger.trace("put_item {}", request);
-
-    auto op = make_shared<put_item_operation>(_proxy, std::move(request));
-    tracing::add_table_name(client_state.get_trace_state(), op->schema()->ks_name(), op->schema()->cf_name());
-    const bool needs_read_before_write = op->request().HasMember("Expected") ||
-            check_needs_read_before_write(op->_condition_expression);
-    if (auto shard = op->shard_for_execute(needs_read_before_write); shard) {
-        _stats.api_operations.put_item--; // uncount on this shard, will be counted in other shard
-        _stats.shard_bounce_for_lwt++;
-        // FIXME: create separate smp_service_group
-        return container().invoke_on(*shard, default_smp_service_group(),
-                [content = std::move(content), cs = client_state.move_to_other_shard()]
-                (executor& e) mutable {
-            return do_with(cs.get(), [&e, content = std::move(content)]
-                                     (service::client_state& client_state) {
-                return e.put_item(client_state, std::move(content));
-            });
-        });
-    }
-    return op->execute(_proxy, client_state, needs_read_before_write, _stats).finally([op, start_time, this] {
-        _stats.api_operations.put_item_latency.add(std::chrono::steady_clock::now() - start_time, _stats.api_operations.put_item_latency._count + 1);
+    return seastar::async([this, &client_state, content = std::move(content)] () mutable -> request_return_type {
```

Comment:
Why did you do this only for put_item() - because this was the context of the original bug report?
Maybe if we decide to use Seastar threads for this (which saddens me, more on this below), we should put the seastar thread in the caller of all the executor functions, instead of in each one separately?

Reply:
It's sad we need to create a Seastar thread, with its rather massive stack, for handling requests which can be highly parallel. I think this is a bad direction to take, but I don't have a simple alternative to suggest :-( Maybe we could at least have a way to have Seastar threads with smaller stacks?

Reply:
As for the first question, I also mentioned it in a comment above: `Naturally, wrapping PutItem in a seastar::thread is just a demo, in order to mitigate stalls all requests should run in a thread.` I just wanted to show with a patch how a threaded request looks like. I'm also not a big fan of wrapping all requests into threads... of course, another way would be to write a future<>-based parser for rapidjson, but that will include lots of copy-pasting, since we'd need to change signatures of many internal rapidjson functions.

Making rapidjson seastar-friendly was actually the theme of one of the student projects, but sadly it wasn't picked by anyone (this year) :)